File: 3827-implement-router/3827-implement-router.py
from collections import deque, defaultdict
from bisect import bisect_left, bisect_right
from typing import List

class Router:
    """
    A class to simulate a network router with a fixed memory limit.
    """

    def __init__(self, memoryLimit: int):
        """
        Initializes the Router with a memory limit and data structures.
        """
        self.memory_limit = memoryLimit
        # Main queue for FIFO logic (forwarding, eviction)
        self.packet_queue = deque()
        # Set for O(1) duplicate detection
        self.packet_set = set()
        # Map for efficient counting: destination -> deque of timestamps
        self.destination_map = defaultdict(deque)

    def addPacket(self, source: int, destination: int, timestamp: int) -> bool:
        """
        Adds a packet to the router. Handles duplicates and memory eviction.
        Returns:
            bool: True if the packet was added, False if it was a duplicate.
        """
        packet = (source, destination, timestamp)
        
        # 1. Check for duplicates using the set for O(1) lookup.
        if packet in self.packet_set:
            return False

        # 2. If the router is full, remove the oldest packet to make space.
        if len(self.packet_queue) == self.memory_limit:
            oldest_packet = self.packet_queue.popleft()
            self.packet_set.remove(oldest_packet)
            
            # Also remove the timestamp from the corresponding destination's deque.
            old_destination = oldest_packet[1]
            self.destination_map[old_destination].popleft()

        # 3. Add the new packet to all data structures.
        self.packet_queue.append(packet)
        self.packet_set.add(packet)
        self.destination_map[destination].append(timestamp)
        
        return True

    def forwardPacket(self) -> List[int]:
        """
        Forwards the next packet in FIFO order.
        Returns:
            List[int]: The forwarded packet [source, dest, time], or [] if empty.
        """
        # 1. Check if the router has any packets.
        if not self.packet_queue:
            return []

        # 2. Get the oldest packet from the front of the queue.
        packet_to_forward = self.packet_queue.popleft()
        
        # 3. Remove it from the helper data structures.
        self.packet_set.remove(packet_to_forward)
        destination = packet_to_forward[1]
        self.destination_map[destination].popleft()
        
        return list(packet_to_forward)

    def getCount(self, destination: int, startTime: int, endTime: int) -> int:
        """
        Counts packets for a destination within an inclusive time range.
        Returns:
            int: The number of matching packets.
        """
        # 1. Check if we have any packets for this destination.
        if destination not in self.destination_map:
            return 0
        
        timestamps = self.destination_map[destination]
        
        # 2. Use binary search to find the number of timestamps in the range.
        # bisect_left finds the first element >= startTime.
        start_index = bisect_left(timestamps, startTime)
        
        # bisect_right finds the first element > endTime.
        end_index = bisect_right(timestamps, endTime)
        
        # The difference is the count of items in [startTime, endTime].
        return end_index - start_index

File: 3827-implement-router/3827-implement-router.java
import java.util.Objects;
import java.util.Queue;
import java.util.LinkedList;
import java.util.Set;
import java.util.HashSet;
import java.util.Map;
import java.util.HashMap;
import java.util.List;
import java.util.ArrayList;

class Router {

    /**
     * Private inner class to represent a packet.
     * Overriding hashCode() and equals() is crucial for the HashSet to detect duplicates correctly.
     */
    private static class Packet {
        int source;
        int destination;
        int timestamp;

        Packet(int source, int destination, int timestamp) {
            this.source = source;
            this.destination = destination;
            this.timestamp = timestamp;
        }

        @Override
        public boolean equals(Object o) {
            if (this == o) return true;
            if (o == null || getClass() != o.getClass()) return false;
            Packet packet = (Packet) o;
            return source == packet.source &&
                   destination == packet.destination &&
                   timestamp == packet.timestamp;
        }

        @Override
        public int hashCode() {
            // Objects.hash provides a convenient way to generate a hash code
            // from multiple fields.
            return Objects.hash(source, destination, timestamp);
        }
    }

    private final int memoryLimit;
    private final Queue<Packet> packetQueue; // For FIFO order
    private final Set<Packet> packetSet;     // For O(1) duplicate detection
    private final Map<Integer, List<Packet>> destinationMap; // For efficient counting

    /**
     * Initializes the Router with a specific memory capacity.
     */
    public Router(int memoryLimit) {
        this.memoryLimit = memoryLimit;
        this.packetQueue = new LinkedList<>();
        this.packetSet = new HashSet<>();
        this.destinationMap = new HashMap<>();
    }
    
    /**
     * Adds a packet to the router.
     * Returns false if the packet is a duplicate.
     * If the router is full, the oldest packet is removed.
     */
    public boolean addPacket(int source, int destination, int timestamp) {
        Packet newPacket = new Packet(source, destination, timestamp);
        
        // Use the HashSet for an efficient duplicate check.
        if (packetSet.contains(newPacket)) {
            return false;
        }

        // If memory limit is reached, evict the oldest packet.
        if (packetQueue.size() == memoryLimit) {
            Packet oldestPacket = packetQueue.poll(); // Dequeue
            
            // Remove the evicted packet from all helper data structures.
            packetSet.remove(oldestPacket);
            destinationMap.get(oldestPacket.destination).remove(0);
        }

        // Add the new packet to all data structures.
        packetQueue.offer(newPacket); // Enqueue
        packetSet.add(newPacket);
        destinationMap.computeIfAbsent(destination, k -> new ArrayList<>()).add(newPacket);
        
        return true;
    }
    
    /**
     * Removes and returns the next packet in FIFO order.
     * Returns an empty array if the router is empty.
     */
    public int[] forwardPacket() {
        if (packetQueue.isEmpty()) {
            return new int[0];
        }

        // Get the oldest packet from the front of the queue.
        Packet packetToForward = packetQueue.poll();
        
        // Remove it from the helper data structures as well.
        packetSet.remove(packetToForward);
        destinationMap.get(packetToForward.destination).remove(0);

        return new int[]{packetToForward.source, packetToForward.destination, packetToForward.timestamp};
    }
    
    /**
     * Counts packets for a specific destination within an inclusive timestamp range.
     */
    public int getCount(int destination, int startTime, int endTime) {
        if (!destinationMap.containsKey(destination)) {
            return 0;
        }

        List<Packet> packets = destinationMap.get(destination);
        
        // Use binary search to efficiently find the packets in the time range.
        int startIndex = findFirstGreaterOrEqual(packets, startTime);
        int endIndex = findFirstGreater(packets, endTime);
        
        return endIndex - startIndex;
    }

    /**
     * Helper method to find the index of the first packet whose timestamp is >= time.
     * This is equivalent to a "lower_bound" search.
     */
    private int findFirstGreaterOrEqual(List<Packet> packets, int time) {
        int low = 0, high = packets.size() - 1;
        int ans = packets.size();
        while (low <= high) {
            int mid = low + (high - low) / 2;
            if (packets.get(mid).timestamp >= time) {
                ans = mid;
                high = mid - 1;
            } else {
                low = mid + 1;
            }
        }
        return ans;
    }
    
    /**
     * Helper method to find the index of the first packet whose timestamp is > time.
     * This is equivalent to an "upper_bound" search.
     */
    private int findFirstGreater(List<Packet> packets, int time) {
        int low = 0, high = packets.size() - 1;
        int ans = packets.size();
        while (low <= high) {
            int mid = low + (high - low) / 2;
            if (packets.get(mid).timestamp > time) {
                ans = mid;
                high = mid - 1;
            } else {
                low = mid + 1;
            }
        }
        return ans;
    }
}